cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.store.preview;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.data2.dataset2.lib.table.leveldb.LevelDBTableCore;
+import co.cask.cdap.data2.dataset2.lib.table.leveldb.LevelDBTableService;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.NamespaceId;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Table responsible for storing the data generated during preview run.
+ */
+public class PreviewTable {
+  private static final DatasetId PREVIEW_TABLE_ID = NamespaceId.SYSTEM.dataset("preview.table");
+  private static final byte[] TRACER = Bytes.toBytes("t");
+  private static final byte[] PROPERTY = Bytes.toBytes("p");
+  private static final byte[] VALUE = Bytes.toBytes("v");
+
+  private final LevelDBTableCore table;
+  private final LevelDBTableService service;
+
+  private final Gson gson = new Gson();
+  private final AtomicLong counter = new AtomicLong(0L);
+
+
+  @Inject
+  PreviewTable(LevelDBTableService service) {
+    try {
+      this.service = service;
+      service.ensureTableExists(PREVIEW_TABLE_ID.getDataset());
+      this.table = new LevelDBTableCore(PREVIEW_TABLE_ID.getDataset(), service);
```

Comment:
Yes. We have different data directory per preview run. So this table would get created in the corresponding directory assigned to the preview run.